### PR TITLE
Return invalid argument for external name services

### PR DESCRIPTION
**Pull request by adleong**
Fixes https://github.com/linkerd/linkerd2/issues/2800#issuecomment-513740498

When the Linkerd proxy sends a query for a Kubernetes external name service to the destination service, the destination service returns `NoEndpoints: exists=false` because an external name service has no endpoints resource.  Due to a change in the proxy's fallback logic, this no longer causes the proxy to fallback to either DNS or SO_ORIG_DST and instead fails the request.  The net effect is that Linkerd fails all requests to external name services.

We change the destination service to instead return `InvalidArgument` for external name services.  This causes the proxy to fallback to SO_ORIG_DST instead of failing the request.

Signed-off-by: Alex Leong <alex@buoyant.io>

### CODE REVIEW

File: controller/api/destination/watcher/k8s.go [hunk 1]
```diff
@@ -24,8 +24,22 @@ type (
 	// Port is a numeric port.
 	Port      = uint32
 	namedPort = intstr.IntOrString
+
+	// InvalidService is n error which indicates that the authority is not a
```
Comment (ihcsim) [inline controller/api/destination/watcher/k8s.go]: s/n/an

File: controller/api/destination/watcher/endpoints_watcher.go [hunk 2]
```diff
@@ -129,6 +129,12 @@ func (ew *EndpointsWatcher) Subscribe(authority string, listener EndpointUpdateL
 	if err != nil {
 		return err
 	}
+
+	svc, _ := ew.k8sAPI.Svc().Lister().Services(id.Namespace).Get(id.Name)
+	if svc != nil && svc.Spec.Type == corev1.ServiceTypeExternalName {
+		return invalidService(authority)
```
Comment (ihcsim) [inline controller/api/destination/watcher/endpoints_watcher.go]: When I set the `destination` container's log level to `debug`, the logs are completely saturated with the `Failed to subscribe` error messages as seen below. I like to think this won't be a problem, as users would only turn on `debug` for temporary debugging purposes, but thought I will bring it to up anyway.

```
time="2019-07-23T03:25:15Z" level=error msg="Failed to subscribe to my-external-nginx.default.svc.cluster.local:80: Invalid k8s service my-external-nginx.default.svc.cluster.local:80" addr=":8086" component=server remote="127.0.0.1:41306"
time="2019-07-23T03:25:15Z" level=debug msg="Get my-external-nginx.default.svc.cluster.local:80" addr=":8086" component=server remote="127.0.0.1:41306"
time="2019-07-23T03:25:15Z" level=error msg="Failed to subscribe to my-external-nginx.default.svc.cluster.local:80: Invalid k8s service my-external-nginx.default.svc.cluster.local:80" addr=":8086" component=server remote="127.0.0.1:41306"
time="2019-07-23T03:25:15Z" level=debug msg="Get my-external-nginx.default.svc.cluster.local:80" addr=":8086" component=server remote="127.0.0.1:41306"
time="2019-07-23T03:25:15Z" level=error msg="Failed to subscribe to my-external-nginx.default.svc.cluster.local:80: Invalid k8s service my-external-nginx.default.svc.cluster.local:80" addr=":8086" component=server remote="127.0.0.1:41306"
time="2019-07-23T03:25:15Z" level=debug msg="Get my-external-nginx.default.svc.cluster.local:80" addr=":8086" component=server remote="127.0.0.1:41306"
time="2019-07-23T03:25:15Z" level=error msg="Failed to subscribe to my-external-nginx.default.svc.cluster.local:80: Invalid k8s service my-external-nginx.default.svc.cluster.local:80" addr=":8086" component=server remote="127.0.0.1:41306"
time="2019-07-23T03:25:15Z" level=debug msg="Get my-external-nginx.default.svc.cluster.local:80" addr=":8086" component=server remote="127.0.0.1:41306"
time="2019-07-23T03:25:15Z" level=error msg="Failed to subscribe to my-external-nginx.default.svc.cluster.local:80: Invalid k8s service my-external-nginx.default.svc.cluster.local:80" addr=":8086" component=server remote="127.0.0.1:41306"
time="2019-07-23T03:25:15Z" level=debug msg="Get my-external-nginx.default.svc.cluster.local:80" addr=":8086" component=server remote="127.0.0.1:41306"
time="2019-07-23T03:25:15Z" level=error msg="Failed to subscribe to my-external-nginx.default.svc.cluster.local:80: Invalid k8s service my-external-nginx.default.svc.cluster.local:80" addr=":8086" component=server remote="127.0.0.1:41306"
time="2019-07-23T03:25:15Z" level=debug msg="Get my-external-nginx.default.svc.cluster.local:80" addr=":8086" component=server remote="127.0.0.1:41306"
time="2019-07-23T03:25:15Z" level=error msg="Failed to subscribe to my-external-nginx.default.svc.cluster.local:80: Invalid k8s service my-external-nginx.default.svc.cluster.local:80" addr=":8086" component=server remote="127.0.0.1:41306"
time="2019-07-23T03:25:15Z" level=debug msg="Get my-external-nginx.default.svc.cluster.local:80" addr=":8086" component=server remote="127.0.0.1:41306"
time="2019-07-23T03:25:15Z" level=error msg="Failed to subscribe to my-external-nginx.default.svc.cluster.local:80: Invalid k8s service my-external-nginx.default.svc.cluster.local:80" addr=":8086" component=server remote="127.0.0.1:41306"
time="2019-07-23T03:25:15Z" level=debug msg="Get my-external-nginx.default.svc.cluster.local:80" addr=":8086" component=server remote="127.0.0.1:41306"
time="2019-07-23T03:25:15Z" level=error msg="Failed to subscribe to my-external-nginx.default.svc.cluster.local:80: Invalid k8s service my-external-nginx.default.svc.cluster.local:80" addr=":8086" component=server remote="127.0.0.1:41306"
time="2019-07-23T03:25:15Z" level=debug msg="Get my-external-nginx.default.svc.cluster.local:80" addr=":8086" component=server remote="127.0.0.1:41306"
time="2019-07-23T03:25:15Z" level=error msg="Failed to subscribe to my-external-nginx.default.svc.cluster.local:80: Invalid k8s service my-external-nginx.default.svc.cluster.local:80" addr=":8086" component=server remote="127.0.0.1:41306"
time="2019-07-23T03:25:15Z" level=debug msg="Get my-external-nginx.default.svc.cluster.local:80" addr=":8086" component=server remote="127.0.0.1:41306"
time="2019-07-23T03:25:15Z" level=error msg="Failed to subscribe to my-external-nginx.default.svc.cluster.local:80: Invalid k8s service my-external-nginx.default.svc.cluster.local:80" addr=":8086" component=server remote="127.0.0.1:41306"
...
```

File: controller/api/destination/watcher/endpoints_watcher.go [hunk 3]
```diff
@@ -129,6 +129,12 @@ func (ew *EndpointsWatcher) Subscribe(authority string, listener EndpointUpdateL
 	if err != nil {
 		return err
 	}
+
+	svc, _ := ew.k8sAPI.Svc().Lister().Services(id.Namespace).Get(id.Name)
```
Comment (alpeb) [inline controller/api/destination/watcher/endpoints_watcher.go]: This same call is repeated in the same code path when not dealing with external services, in `newPortPublisher()`. From my understanding there's a shared informer underneath that caches that, but given the informer is instantiated again that caching is not happening, right? If so, you could consider shuffling things a bit to have that call happen only once.

**Comment by adleong** [comment]
This will conflict significantly with https://github.com/linkerd/linkerd2/pull/3113.  I will wait until that PR merges before resuming work on this.

File: controller/api/destination/watcher/endpoints_watcher.go (hunk 3 above)
Comment (adleong) [reply controller/api/destination/watcher/endpoints_watcher.go]: I don't think the informer is being instantiated again.  there should only be one informer instance, I believe

File: controller/api/destination/watcher/endpoints_watcher.go (hunk 2 above)
Comment (adleong) [reply controller/api/destination/watcher/endpoints_watcher.go]: Looks like half of those are logged at error level, which is problematic...

File: controller/api/destination/watcher/endpoints_watcher.go (hunk 3 above)
Comment (alpeb) [reply controller/api/destination/watcher/endpoints_watcher.go]: My bad, the informer is indeed kept in `API.svc` and it's reused.

**Pull request merged**